DS-712 Remove twig whitespace added around tooltips

MarcinMr opened the pull request:
## Jira

https://pegadigitalit.atlassian.net/browse/DS-712

## Summary

Whitespace around the tooltip trigger was removed.

## Details

Whitespace was removed using CSS properties. `Spaceless` didn't want to work as mentioned in the Jira description.

## How to test

- Review code
- Check out this branch locally and see the new test page: `999-tests --> Tooltip Whitespace` (`/pattern-lab/?p=tests-tooltip-whitespace`)
- Check if there is no more whitespace around the tooltip trigger.

colbytcook reviewed (approved):
Confirmed this fix works on Safari (OS and iOS), Firefox, Chrome and Edge (Mac)

mikemai2awesome commented on docs-site/src/pages/pattern-lab/_patterns/40-components/tooltip/15-tooltip-inverted.twig:
```diff
@@ -1,7 +1,7 @@
   {# Start component specific code #}
   {% include '@bolt-components-tooltip/tooltip.twig' with {
-    trigger: 'this example',
+    trigger: 'this example.',
```
Why did this doc change? The period should be outside of the tooltip.

MarcinMr replied:
<img width="178" alt="dot-space" src="https://user-images.githubusercontent.com/85670248/162912638-bcc76a91-14d8-4bde-9b1a-543ef9dd75d6.png">

<img width="174" alt="dot-space-2" src="https://user-images.githubusercontent.com/85670248/162912653-32c66fbe-d65f-4bfd-ac69-1415704e6d83.png">

I've changed it because when the dot is outside the trigger there is still space between the text and the dot, like in the examples above.

Here `/pattern-lab/?p=tests-tooltip-whitespace` I updated examples of tooltip triggers.

mikemai2awesome replied:
That is exactly the problem. We need to solve the undesired white-space in between the text and the period. Not to change the doc to hide the problem.

MarcinMr replied:
I'm not sure the tooltip itself and the `display: inline-block;` cause the issue.

Here is an example of `display: inline;` We can see that the dotted line stretches from the start of the string to the end of the dot.

<img width="113" alt="display-inline" src="https://user-images.githubusercontent.com/85670248/163120235-17265616-172b-45f2-898c-ad78ecd72664.png">

Here is the `display: inline-block;` and the dotted line stretches from the start of the string to the end of the string. So this is how it should look like but instead of this whitespace before the dot.

<img width="112" alt="display-inline-block" src="https://user-images.githubusercontent.com/85670248/163120396-1b415906-600a-452b-a61a-cf0e2ea61f3e.png">

Is the tooltip itself causing the issue and not the abandoned dot? In the dev tools if you check the "this example" string, whitespace is only before that string. After the string, there is no whitespace. But the dot has whitespace in front of it and back of it.

<img width="224" alt="tooltip whitespace" src="https://user-images.githubusercontent.com/85670248/163121021-dba3d931-3038-4d52-8d6f-301bfb564e6b.png">

Look at this record. Changing the "this example" string doesn't cause anything, but deleting the whitespace from the dot fix the issue.

https://user-images.githubusercontent.com/85670248/163108972-a8056a6f-df21-4810-82cf-58ac260dea5c.mov

mikemai2awesome replied:
This is a problem we need to solve, we have used `spaceless` elsewhere.

MarcinMr replied:
Filter `|spaceless` works in demo pages but with a combination of `display: inline-block`. that filter doesn't work with `display: inline`.

mikemai2awesome commented on docs-site/src/pages/pattern-lab/_patterns/999-tests/tooltip/20-tooltip-whitespace.twig:
```diff
@@ -0,0 +1,9 @@
+{% set tooltip %}
```
This test is not sufficient. To test if the tooltip is truly inline without undesired white space, you'd need to layout all the scenarios such as:

This is a paragraph of text, a tooltip can appear anywhere in here, in the middle of the sentence **_like this_** or at the end of the sentence before a punctuation **_like this_**.

$**_A tooltip_** can also begin a sentence where a special character is used before it.

mikemai2awesome replied:
Good. Now this test perfectly demonstrates the undesired white space, which is a side effect of `inline-block`.

![whitespace](https://user-images.githubusercontent.com/3027663/163031327-3108f939-9a70-416f-ad95-0bf41e356cf4.jpg)

MarcinMr commented:
@danielamorse 

I used a few `<spaceless>` tag combinations and some indentation changing and that works but only for the dot at the end of the tooltip. There is still space in front of the tooltip when we use a special character.

It looks like this in the dev tools, there is no space around tooltip and around the dot, but still exist near special character in front of the tooltip string:
<img width="486" alt="special sign whitespace" src="https://user-images.githubusercontent.com/85670248/164481838-430d9350-8c9f-4864-9585-eb4f47fde32c.png">

I tried to remove everything and reduce the code to one line for testing purposes like this:
<img width="605" alt="spaceless one line doesnt work" src="https://user-images.githubusercontent.com/85670248/164482434-de891dc2-826b-4b5c-b7a1-8d3aef382dfa.png">
I also tried adding `|spaceless` filter to the above code like this: `{{ trigger|spaceless }}`, but didn't help.

Tried also put there basic string:
<img width="731" alt="whitespace test string" src="https://user-images.githubusercontent.com/85670248/164482997-a3d3015a-3334-42e7-aa5d-0433baf6fa04.png">
An the filter removes spaces around the tooltip but there is still whitespace near the special character:
<img width="380" alt="white-space issue" src="https://user-images.githubusercontent.com/85670248/164483159-891f2cd4-c8a9-4f53-a2dc-17d0cfc41d68.png">

I tried also with `|trim` but didn't help :(

The only thing that worked was using the `|spacless` filter but in the demo pages, with the combination of `display: inline-block` in CSS.
<img width="439" alt="display-inline-block-spaceless-demo-doc" src="https://user-images.githubusercontent.com/85670248/164483872-49b60df2-21b1-40c8-abf6-7e449b0e1c74.png">

I don't understand why the `spaceless` filter works on the demo pages but when I added it in the template and reduce code to one line to be sure there are no spaces, that didn't help.

I'm still working on that and will try also tomorrow. Dan, do you have any more ideas for that?

danielamorse reviewed (changes requested):
@MarcinMr you've actually solved the issue we care about. There is no longer extra space _inside_ the tooltip [here](https://feature-ds-712-remove-twig-whitespace-add.boltdesignsystem.com/pattern-lab/?p=components-tooltip).

In the other examples you mention, the extra space is getting added _outside_ the tooltip when the variable is assigned.

```
{% set tooltip %}// This line-break is adding whitespace
  {# Start component specific code #}
  {% include '@bolt-components-tooltip/tooltip.twig' with {
    trigger: 'this example',
    content: 'This is an inverted tooltip.',
    inverted: true
  } only %}
  {# End component specific code #}
{% endset %}
```

You'd have to do this:

```
{% set tooltip %}{% spaceless %}
  {# Start component specific code #}
  {% include '@bolt-components-tooltip/tooltip.twig' with {
    trigger: 'this example',
    content: 'This is an inverted tooltip.',
    inverted: true
  } only %}
  {# End component specific code #}
{% endspaceless %}{% endset %}
```

There is nothing you can do to trim that space. You must use `{% spaceless %}` when you assign the variable,  `|spaceless` when you print the variable, or simply don't add a line-break in the first place. _That_ space is out of our control (aside from [hacks we don't want to employ](https://css-tricks.com/fighting-the-space-between-inline-block-elements/) here). We must manually trim it in our demos and we should probably add a note saying something like: "When adding an inline tooltip be sure to trim any unwanted space around the element."

Thanks for adding the test page. Would you please update that test page and the other demo you modified to use `{% spaceless %}` and add a comment to the basic Tooltip demo explaining how you must trim the whitespace around the tooltip?

If you have any questions we can chat tomorrow 🙏

danielamorse commented:
@MarcinMr I also noticed that space is added by the top of the Tooltip Twig template [where we assign variables](https://github.com/boltdesignsystem/bolt/blob/bee535c7e333201fea6a0772fbcc831f90b196c2/packages/components/bolt-tooltip/src/tooltip.twig#L1-L58). To eliminate that space we'd either have to do something like this:

```
{% spaceless %}
{% set schema = bolt.data.components["@bolt-components-tooltip"].schema %}
...
```

Or just recommend that anytime you use a Tooltip inline that you must use spaceless in your variable assignment or variable output.

What do you think? Let's talk more tomorrow morning.

MarcinMr commented:
> @MarcinMr I also noticed that space is added by the top of the Tooltip Twig template [where we assign variables](https://github.com/boltdesignsystem/bolt/blob/bee535c7e333201fea6a0772fbcc831f90b196c2/packages/components/bolt-tooltip/src/tooltip.twig#L1-L58). To eliminate that space we'd either have to do something like this:
> 
> ```
> {% spaceless %}
> {% set schema = bolt.data.components["@bolt-components-tooltip"].schema %}
> ...
> ```
> 
> Or just recommend that anytime you use a Tooltip inline that you must use spaceless in your variable assignment or variable output.
> 
> What do you think? Let's talk more tomorrow morning.

I added your recommended solution but we still need to add a `spaceless` or `|spaceless` filter in demos, when we have a special character at the beginning of a tooltip.

> @MarcinMr you've actually solved the issue we care about. There is no longer extra space _inside_ the tooltip [here](https://feature-ds-712-remove-twig-whitespace-add.boltdesignsystem.com/pattern-lab/?p=components-tooltip).
> 
> In the other examples you mention, the extra space is getting added _outside_ the tooltip when the variable is assigned.
> 
> ```
> {% set tooltip %}// This line-break is adding whitespace
>   {# Start component specific code #}
>   {% include '@bolt-components-tooltip/tooltip.twig' with {
>     trigger: 'this example',
>     content: 'This is an inverted tooltip.',
>     inverted: true
>   } only %}
>   {# End component specific code #}
> {% endset %}
> ```
> 
> You'd have to do this:
> 
> ```
> {% set tooltip %}{% spaceless %}
>   {# Start component specific code #}
>   {% include '@bolt-components-tooltip/tooltip.twig' with {
>     trigger: 'this example',
>     content: 'This is an inverted tooltip.',
>     inverted: true
>   } only %}
>   {# End component specific code #}
> {% endspaceless %}{% endset %}
> ```
> 
> There is nothing you can do to trim that space. You must use `{% spaceless %}` when you assign the variable, `|spaceless` when you print the variable, or simply don't add a line-break in the first place. _That_ space is out of our control (aside from [hacks we don't want to employ](https://css-tricks.com/fighting-the-space-between-inline-block-elements/) here). We must manually trim it in our demos and we should probably add a note saying something like: "When adding an inline tooltip be sure to trim any unwanted space around the element."
> 
> Thanks for adding the test page. Would you please update that test page and the other demo you modified to use `{% spaceless %}` and add a comment to the basic Tooltip demo explaining how you must trim the whitespace around the tooltip?
> 
> If you have any questions we can chat tomorrow 🙏

@danielamorse thanks for explaining that. I also noticed that some of the comments could cause problems with whitespace, so I deleted them. I added a note in the tooltip docs and also updated the test page. Let me know if that's sufficient and if you spot anything there that is causing any issues related to whitespace.

danielamorse reviewed (approved):
Everything looks good now. Nice job on the docs!
